### PR TITLE
feat(connlib): smoothly migrate relayed connections

### LINKED ISSUE #4548: Refresh relays messages
Not necessarily the cause behind #4521 but might partially be.

Right now we get the relays credentials after we generate a connection intent, we create a new connection and we use the given TURN servers with the given credentials throughout the lifetime of the connection.

However, when there is a redploy the relays can get restarted or deleted making those credentials or whole server invalid.

The portal already have a mechanism implemented where a new relay is deployed before taking down the old one, with a new message for the gateway and client we can greatly reduce or even prevent any downtime.

With the portal broadcasting the new servers for the given connections and with a message for invalidating the old server, we can inform this to snownet and invalidate the corresponding candidates and generate new ones.

Right now the highest priority would be just to be able to inform the new TURN candidates(new servers) since with that the old ones would just timeout, without this the peer connection would be terminated invalidating all DNS mappings which would take a long time to self-heal.

### PR DESCRIPTION
**Pull request by thomaseizinger**
Whenever we receive a `relays_presence` message from the portal, we invalidate the candidates of all now disconnected relays and make allocations on the new ones. This triggers signalling of new candidates to the remote party and migrates the connection to the newly nominated socket.

This still relies on #4613 until we have #4634.

Resolves: #4548.

### CODE REVIEW

**Comment by thomaseizinger** [comment]
This is only waiting on https://github.com/firezone/firezone/pull/4567 and https://github.com/firezone/firezone/pull/4596#discussion_r1564792450.

**Comment by thomaseizinger** [comment]
I want to time-box writing a unit-test for this.

File: rust/Cargo.lock
Comment (thomaseizinger) [inline rust/Cargo.lock]: Needed to patch str0m: https://github.com/algesten/str0m/pull/497

File: rust/connlib/snownet/src/node.rs
Comment (thomaseizinger) [inline rust/connlib/snownet/src/node.rs:1746]: Newly added function: https://github.com/algesten/str0m/pull/497.

**Comment by thomaseizinger** [comment]
cc @conectado Ready for review.

File: scripts/tests/relayed-curl-relay-restart.sh
Comment (jamilbk) [inline scripts/tests/relayed-curl-relay-restart.sh]:
```suggestion
RELAY_2_PUBLIC_IP4_ADDR="172.28.0.202" docker compose up -d relay-2
```

**Review by jamilbk — CHANGES_REQUESTED** [review]
Typo

File: docker-compose.yml
Comment (jamilbk) [inline docker-compose.yml]:
```suggestion
```

Let's leave `relay-1`'s port specifiers in case we need to test another client with the local dev setup, but these can be removed

File: docker-compose.yml
Comment (jamilbk) [inline docker-compose.yml]:
```suggestion
```

**Review by jamilbk — CHANGES_REQUESTED** [review]

**Comment by jamilbk** [comment]
@thomaseizinger ~~Merging my suggestions to see if this PR passes tests.~~

Edit: saw you already applied them

**Review by jamilbk — COMMENTED** [review]
LGTM. We'll want to get this onto `staging` for testing with newly-built clients.

Deferring approval to @conectado

**Review by jamilbk — DISMISSED** [review]

**Comment by conectado** [comment]
We should test this with #4569

**Review by conectado — APPROVED** [review]
LGTM!

**Comment by thomaseizinger** [comment]
> We should test this with #4569

@conectado Does your approval after this mean you tested it?

**Comment by jamilbk** [comment]
> > We should test this with #4569
> 
> @conectado Does your approval after this mean you tested it?

I updated the base of that PR to this one, and the browser test is failing unfortunately there

**Comment by thomaseizinger** [comment]
Several CI failures are expected here:

- Perf-tests for HEAD fail because I've renamed the compose services so they can't find `relay-1` and `relay-2`
- Comp tests for DNS fail because we restart the relay using a different IP, thus not having any available relays (and the released versions do not yet parse the `relays_presence` message). The DNS restart test passes for this branch though!

**Comment by thomaseizinger** [comment]
@conectado Ready for re-review, not much has changed though. Some CI failures are expected as described in https://github.com/firezone/firezone/pull/4568#issuecomment-2062818658.

**Comment by conectado** [comment]
> > We should test this with #4569
> 
> @conectado Does your approval after this mean you tested it?

I haven't tested it, do you think it needs additional manual testing?

**Comment by conectado** [comment]
> > > We should test this with #4569
> > 
> > 
> > @conectado Does your approval after this mean you tested it?
> 
> I updated the base of that PR to this one, and the browser test is failing unfortunately there

This shouldn't be failing, for us to merge though

**Comment by jamilbk** [comment]
> This shouldn't be failing, for us to merge though

I think the intent was that it should now be passing assuming that PR is based off this one.

**Comment by jamilbk** [comment]
> Comp tests for DNS fail because we restart the relay using a different IP, thus not having any available relays (and the released versions do not yet parse the relays_presence message). The DNS restart test passes for this branch though!

~~Hm, why what that cause DNS to fail? That means that deploying this change will break everyone's DNS resources whenever we deploy, requiring them to restart their clients, correct (until they can upgrade their clients/gateways)? This can't go out until the macOS clients is approved then.~~

~~Is the API breakage required?~~

Scratch that, I misunderstood. This test only caught an existing issue that's already there.

LGTM, will wait for @conectado's approval before merging.

**Comment by thomaseizinger** [comment]
Just resolved the conflicts, is anything missing from an approval here @conectado ?

**Review by conectado — APPROVED** [review]
Looks good to me, I could swear I had approved this

**Comment by thomaseizinger** [comment]
@jamilbk what are we doing about the failed compat test? Can we mark that as non-required until the next release? Otherwise all PRs following this won't be able to merge.

**Comment by thomaseizinger** [comment]
> Looks good to me, I could swear I had approved this

You did approve a much earlier version but lots of rebasing etc happen since :)

**Comment by jamilbk** [comment]
> @jamilbk what are we doing about the failed compat test? Can we mark that as non-required until the next release? Otherwise all PRs following this won't be able to merge.

Yeah, or cut a new release after this is merged. I'll probably do the former and some more testing with all the clients later tonight so we can get ahead of any issues for Monday.

**Comment by jamilbk** [comment]
Merging to do some testing on staging with new client builds over the weekend.

**Pull request merged**
